src/arch/MovieTexture/MovieTexture_FFMpeg.cpp

```diff
@@ -164,93 +166,128 @@ float MovieDecoder_FFMpeg::GetTimestamp() const
 	}
 
 	// In a logical situation, this means that display is outpacing decoding.
-	if (display_frame_num_ >= static_cast<int>(frame_buffer_.size())) {
+	if (display_frame_num_ >= static_cast<int>(packet_buffer_.size())) {
+		return 0;
+	}
+
+	PacketHolder* packet = packet_buffer_[display_frame_num_].get();
+
+	// Sanity check.
+	if (packet == nullptr) {
 		return 0;
 	}
-	return frame_buffer_[display_frame_num_]->frame_timestamp;
+
+	std::lock_guard<std::mutex> lock(packet->lock);
+	return packet->frame_timestamp;
 }
 
 bool MovieDecoder_FFMpeg::IsCurrentFrameReady() {
-	// We're displaying faster than decoding. Do not even try to display the frame.
-	if (display_frame_num_ >= static_cast<int>(frame_buffer_.size())) {
+	// We're displaying faster than decoding. Do not even try to display the
+	// frame.
+	if (display_frame_num_ >= static_cast<int>(packet_buffer_.size())) {
 		return false;
 	}
-	// If the whole movie is decoded, then the frame is definitely ready.
-	if (end_of_file_) {
-		return true;
-	}
 
+	FrameHolder* frame = frame_buffer_[(display_frame_num_ + offset_) % frame_buffer_.size()].get();
+	// To make sure the frame doesn't change from under us.
+	std::lock_guard<std::mutex> lock(frame->lock);
 
-	std::lock_guard<std::mutex> lock(frame_buffer_[display_frame_num_]->lock);
-	if (frame_buffer_[display_frame_num_]->skip) {
-		LOG->Info("Frame %i not decoded, skipping...", display_frame_num_);
-		return true;
+	// In terms of how the sliding window works, displayed and ready are
+	// opposites. If the frame hasn't been displayed, it's ready. If it has
+	// been displayed, then it hasn't been overwritten and reset yet.
+	if (frame->displayed) {
+		LOG->Info("Frame %i not decoded, total frames: %i", display_frame_num_, total_frames_);
 	}
-	if (!frame_buffer_[display_frame_num_]->decoded) {
-		LOG->Info("Frame %i not decoded and was not skipped, total frames: %i", display_frame_num_, total_frames_);
+	return !frame->displayed;
+}
+
+int MovieDecoder_FFMpeg::HandleNextPacket() {
+	// If the decoder hit the end of the file, then that means the packet
+	// buffer is complete.
+	if (!end_of_file_) {
+		// Add in a new FrameBuffer entry, and lock it immediately.
+		packet_buffer_.emplace_back(std::make_unique<PacketHolder>());
+		std::unique_lock<std::mutex> lock(packet_buffer_.back()->lock);
+		int status = SendPacketToBuffer();
+		if (status < 0) {
+			lock.unlock();
+			return status;
+		}
```

Comment:
I would think it'd be preferable to use `std::unique_lock` to automatically release the lock when it goes out of scope. Is that a possibility?

Reply:
The problem is that the lock belongs to the `PacketHolder`. So when we `pop_back` below, the lock is deleted. If we use a `std::unique_lock`, then the program will try and release the already deleted lock.

If we had some way to `peek` the next packet/frame, that could simplify this, but it feels fine for now.